### ISSUE TITLE
Incompatibility with the last Intel NEO drivers

### ISSUE THREAD

**Issue by Nyrio** [CNugteren/CLBlast#355]
I am using CLBlast on a bunch of Ubuntu 18 machines. I am using a manual installation, not the Ubuntu packages.

Everything was working fine for weeks, then I have updated the NEO drivers and it's now crashing with the following error (same for `Axpy`, `Gemv`, `Gemm`, etc):
```
OpenCL compiler error/warning:
In file included from <built-in>:2:
CTHeader.h:6913:7: error: use of type 'intel_sub_group_avc_mce_payload_t' requires cl_intel_device_side_avc_motion_estimation extension to be enabled
      intel_sub_group_avc_mce_payload_t payload );
```
(the error appears many times with different line and column numbers)

Previous version:
```
Device Version    OpenCL 2.1 NEO
Driver Version    19.05.12254
Device OpenCL C Version    OpenCL C 2.0
```

Current version:
```
Device Version    OpenCL 2.1 NEO 
Driver Version    19.16.12873
Device OpenCL C Version    OpenCL C 2.0
```

Also, the extension `cl_intel_device_side_avc_motion_estimation` seems to be there (though reported as missing):
```
Device Extensions cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_fp16 cl_khr_depth_images cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_icd cl_khr_image2d_from_buffer cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_intel_subgroups cl_intel_required_subgroup_size cl_intel_subgroups_short cl_khr_spir cl_intel_accelerator cl_intel_media_block_io cl_intel_driver_diagnostics cl_intel_device_side_avc_motion_estimation cl_khr_priority_hints cl_khr_throttle_hints cl_khr_create_command_queue cl_khr_fp64 cl_khr_subgroups cl_khr_il_program cl_intel_spirv_device_side_avc_motion_estimation cl_intel_spirv_media_block_io cl_intel_spirv_subgroups cl_khr_spirv_no_integer_wrap_decoration cl_khr_mipmap_image cl_khr_mipmap_image_writes cl_intel_planar_yuv cl_intel_packed_yuv cl_intel_motion_estimation cl_intel_advanced_motion_estimation cl_intel_va_api_media_sharing 
```

If there is a way to disable subgroups while waiting for this issue to be addressed, please let me know.

**Comment by CNugteren** [comment]
Thanks for reporting. As a workaround you can just remove lines 64 and 65 of [utilities/compile.cpp](https://github.com/CNugteren/CLBlast/blob/master/src/utilities/compile.cpp#L64).

To solve the issue, I guess we just need to add the following to the OpenCL kernel code:
```
#pragma OPENCL EXTENSION cl_intel_device_side_avc_motion_estimation
```
as well as a check in the host code to make sure this extension is available.

Perhaps someone from the Intel NEO team can confirm that this is indeed now needed and not a regression in Intel NEO itself? Perhaps @JacekDanecki can give some advice here?

**Comment by JacekDanecki** [comment]
@CNugteren you need to add in kernel code:
```
#pragma OPENCL EXTENSION cl_intel_device_side_avc_motion_estimation : enable
```
As [IGC compiler](https://github.com/intel/intel-graphics-compiler) switched from llvm 4 to llvm 7, newer clang requires this pragma.

**Comment by CNugteren** [comment]
Thanks for the help. And will this then break on older NEO drivers, or will that still work?

**Comment by JacekDanecki** [comment]
Per OpenCL spec a
```
#pragma OPENCL EXTENSION extension_name : enable
```
is required if we want to use a language extension. With Clang based on LLVM 4.0 this requirement was omitted e.g. for VME extension. 
So, it should not break older NEO versions.

**Comment by CNugteren** [comment]
I've made the PR for the suggested fix: https://github.com/CNugteren/CLBlast/pull/357

@Nyrio: could you test with that branch (CLBlast-355-intel-shuffle-extension-fix) and let us know if it works.

@JacekDanecki: I still find it a bit of a weird issue. To me the name `cl_intel_device_side_avc_motion_estimation` seems unrelated to the shuffling instructions I'm using (`get_sub_group_local_id` and `intel_sub_group_shuffle`), and it also isn't mentioned at all in the [extension document](https://www.khronos.org/registry/OpenCL/extensions/intel/cl_intel_subgroups.html).

**Comment by Nyrio** [comment]
@CNugteren : unfortunately I still have the issue with this branch:
```
OpenCL compiler error/warning:
In file included from <built-in>:2:
CTHeader.h:6913:7: error: use of type 'intel_sub_group_avc_mce_payload_t' requires cl_intel_device_side_avc_motion_estimation extension to be enabled
      intel_sub_group_avc_mce_payload_t payload );
```

**Comment by Nyrio** [comment]
Oh, I also have:
```
1:2179:28: warning: unsupported OpenCL extension 'cl_intel_device_side_avc_motion_estimation' - ignoring
  #pragma OPENCL EXTENSION cl_intel_device_side_avc_motion_estimation: enable
```

**Comment by JacekDanecki** [comment]
@CNugteren to use cl_intel_subgroups extension use this pragma
```
#pragma OPENCL EXTENSION  cl_intel_subgroups : enable
```
For each extension you need to use pragma as
```
#pragma OPENCL EXTENSION extension_name : enable
```

**Comment by CNugteren** [comment]
Thanks for testing. I'll update the code as soon as possible, but it seems there might be two bugs in NEO then:
* The original error message is wrong?
* The unneeded motion estimation extension doesnt work?

**Comment by CNugteren** [comment]
@Nyrio: The change suggested by @JacekDanecki is now implemented, could you try again?

**Comment by Nyrio** [comment]
Unfortunately, the same problem with that at the end of the logs:
```
1:2179:28: warning: unsupported OpenCL extension 'cl_intel_subgroups' - ignoring
  #pragma OPENCL EXTENSION cl_intel_subgroups: enable
```

**Comment by JacekDanecki** [comment]
Can you create issue in [Neo](https://github.com/intel/compute-runtime) repository, with steps how to reproduce this, so I can try to debug it on my side?

**Comment by Nyrio** [comment]
@CNugteren from what has been said on https://github.com/intel/compute-runtime/issues/167, it seems that `cl_intel_subgroups` requires OpenCL 1.2 and CLBlast uses 1.1.

Could you make a patch with `-cl-std=CL1.1` becoming `-cl-std=CL1.2`?

**Comment by CNugteren** [comment]
OK, thanks for reporting the issue there. I've just pushed code to that branch that enables 1.2 for the kernels in case that extension is present. Could you let me know if it works?

**Comment by Nyrio** [comment]
@CNugteren have you pushed your changes to that branch? I don't see the change on #357 and the last commit on CLBlast-355-intel-shuffle-extension-fix dates was on May 8th

**Comment by CNugteren** [comment]
Oops, sorry. What about now?

**Comment by Nyrio** [comment]
It's working! \o/
Thanks

**Comment by Nyrio** [comment]
We can close the issue when the patch is merged

**Comment by CNugteren** [comment]
All merged in :-)

**Comment by PawelJurek** [comment]
Hi, just for your information: 
the confusing error message when `-cl-std=CL1.1` was used has been fixed in [igc-1.0.7 release](https://github.com/intel/intel-graphics-compiler/releases/tag/igc-1.0.7).
Thank you for reporting this.
